technology_and_tooling/testing/testable_code_fixtures.md

```diff
@@ -0,0 +1,799 @@
+
+    ...
+~~~
+
+:::
+::::
+
+In our tests for the `Trial` class, we have to initialise the class using a CSV file in order to create an instance, even if we do not use that particular data in our tests.  How can we simplify this? One thing that can be changed is the `__init__` method, if we just needed the data as an argument, rather than the path to a CSV file, that would make testing easier. After this change, a separate method is going to be needed to allow creating a `Trial` from a CSV filepath, this can be achieved using a class method. 
+
+
+::::challenge{id=load-from-csv title="Refactor the `Trial` class."}
+
+As described above, refactor the `--init__` method of the `Trial` class to take `data` and `id` as arguments. Then create a class method called `from_csv` that accepts a filepath to a CSV file and returns an instance of the `Trial` class with the data loaded.
+
+:::solution
+Here is the first section of our adjusted object code:
+~~~python
+class Trial:
+    def __init__(self, data, id):
+        self.data = data
+        self.id = id
+
+   @classmethod
+    def from_csv(cls, filename, id):
+        """
+        Class method to create a Trial instance from data in a CSV file.
+
+        Parameters:
+        filename (str): The file path of the CSV file to read.
+        id (str): The id to assign to the Trial instance.
+
+        Returns:
+        Trial: A Trial instance with the data and id from the CSV file.
+        """
+        data = cls.load_csv(filename)
+        return cls(data, id)
+
+    @staticmethod
+    def load_csv(filename):
+        """Load a Numpy array from a CSV
+
+        :param filename: Filename of CSV to load
+        """
+        return np.loadtxt(fname=filename, delimiter=',')
+
+    ...
+~~~
+
+:::
+::::
+
+Now, a `Trial` object can be created in two ways:
+
+~~~python
+import numpy as np
+from inflammation.models import Trial
+
+filename = "inflammation-01.csv"
+data = np.loadtxt(fname=filename, delimiter=',')
+
+trial_group_01 = Trial(data, "Group01")
+trial_group_02 = Trial.from_csv("inflammation-02.csv", "Group02")
+~~~
+
+For our tests, we no longer need a CSV file in order to ensure that the statistical methods from the class give the expected results and we can replace our `trial_instance` fixture:
+
+~~~python
+@pytest.fixture()
+def trial_instance():
+    return Trial(np.array([[0, 0],[0, 0]]), 1)
+~~~
+
+Alternatively, we can create objects within test methods, if we prefer to do things that way:
+
+~~~python
+class TestTrial:
+    def test_daily_mean_zeros(self):
+        """Test that mean function works for an array of zeros."""
+        trial_instance = Trial(np.array([[0, 0],[0, 0],[0, 0]]), "Test")
+        test_result = np.array([0, 0])
+
+        # Need to use Numpy testing functions to compare arrays
+        npt.assert_array_equal(trial_instance.daily_mean(), test_result)
+
+    ...
+~~~
+
+### Using a database rather than CSV files
```

Comment:
I feel this section is getting quite long, and the database stuff seems like it could be a separate section. Pre-database focusses more on designing testable code, and post-database is more on fixtures. So the first section could be "Testable Code" and the second: "More about Fixtures", or just "Fixtures"

Reply:
after discussion I think we're just going to leave this as is?

Reply:
Hi @martinjrobins , yes let's leave it as is.